ocaml/xapi/xapi_pool.ml

```diff
@@ -315,10 +315,29 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
 
     match my_backend' with
     | Network_interface.Openvswitch ->
-      let my_controller = Db.Pool.get_vswitch_controller ~__context ~self:my_pool in
-      let controller = Client.Pool.get_vswitch_controller ~rpc ~session_id ~self:pool in
-      if my_controller <> controller && my_controller <> "" then
-        raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["vswitch controller address differs"]))
+      let my_legacy_controller = Db.Pool.get_vswitch_controller ~__context ~self:my_pool in
+      let remote_legacy_controller = Client.Pool.get_vswitch_controller ~rpc ~session_id ~self:pool in
+      if my_legacy_controller <> remote_legacy_controller && my_legacy_controller <> "" then
+        raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["vswitch controller address differs"]));
+
+      if List.length (Db.SDN_controller.get_all ~__context) > 0 then begin
```

Comment:
Better use a `match` here, e.g.:
```
let remote_sdn_controllers = Client.SDN_controller.get_all ~rpc ~session_id in
let my_sdn_controllers = Db.SDN_controller.get_all ~__context in
(* We assume that each pool has _at most_ one SDN controller *)
match remote_sdn_controllers, my_sdn_controllers with
| [], [] -> ()
| remote_sdn_controller :: _, my_sdn_controller :: _ ->
  (* check that protocol/address/port are identical *)
| _ ->
  raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["SDN controller differs"]))
```
The last case handles both the cases "I have a controller, but the remote doesn't" as well as "the remote has a controller, but I don't". I don't think your code does that.